File: ios/LibVlcPlayerModule.swift
import ExpoModulesCore

private let bufferingEvent = "onBuffering"
private let playingEvent = "onPlaying"
private let pausedEvent = "onPaused"
private let stoppedEvent = "onStopped"
private let endReachedEvent = "onEndReached"
private let encounteredErrorEvent = "onEncounteredError"
private let timeChangedEvent = "onTimeChanged"
private let positionChangedEvent = "onPositionChanged"
private let esAddedEvent = "onESAdded"
private let dialogDisplayEvent = "onDialogDisplay"
private let firstPlayEvent = "onFirstPlay"
private let backgroundEvent = "onBackground"

let playerEvents = [
    bufferingEvent,
    playingEvent,
    pausedEvent,
    stoppedEvent,
    endReachedEvent,
    encounteredErrorEvent,
    timeChangedEvent,
    positionChangedEvent,
    esAddedEvent,
    dialogDisplayEvent,
    firstPlayEvent,
    backgroundEvent,
]

public class LibVlcPlayerModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoLibVlcPlayer")

        OnDestroy {
            MediaPlayerManager.shared.onModuleDestroyed()
        }

        View(LibVlcPlayerView.self) {
            Events(playerEvents)

            Prop("source") { (view: LibVlcPlayerView, source: String?) in
                view.source = source
            }

            Prop("options") { (view: LibVlcPlayerView, options: [String]?) in
                view.options = options ?? [String]()
            }

            Prop("tracks") { (view: LibVlcPlayerView, tracks: Tracks?) in
                view.tracks = tracks
            }

            Prop("slaves") { (view: LibVlcPlayerView, slaves: [Slave]?) in
                view.slaves = slaves ?? [Slave]()
            }

            Prop("scale") { (view: LibVlcPlayerView, scale: Float?) in
                view.scale = scale ?? defaultPlayerScale
            }

            Prop("aspectRatio") { (view: LibVlcPlayerView, aspectRatio: String?) in
                view.aspectRatio = aspectRatio
            }

            Prop("contentFit") { (view: LibVlcPlayerView, contentFit: String?) in
                if let contentFit = contentFit, let mode = ResizeMode(rawValue: contentFit) {
                    view.contentFit = mode
                } else {
                    view.contentFit = .contain
                }
            }

            Prop("rate") { (view: LibVlcPlayerView, rate: Float?) in
                view.rate = rate ?? defaultPlayerRate
            }

            Prop("time") { (view: LibVlcPlayerView, time: Int?) in
                view.time = time ?? defaultPlayerTime
            }

            Prop("volume") { (view: LibVlcPlayerView, volume: Int?) in
                view.volume = volume ?? maxPlayerVolume
            }

            Prop("mute") { (view: LibVlcPlayerView, mute: Bool?) in
                view.mute = mute ?? false
            }

            Prop("audioMixingMode") { (view: LibVlcPlayerView, audioMixingMode: AudioMixingMode?) in
                view.audioMixingMode = audioMixingMode ?? .auto
            }

            Prop("repeat") { (view: LibVlcPlayerView, shouldRepeat: Bool?) in
                view.shouldRepeat = shouldRepeat ?? false
            }

            Prop("playInBackground") { (view: LibVlcPlayerView, playInBackground: Bool?) in
                view.playInBackground = playInBackground ?? false
            }

            Prop("autoplay") { (view: LibVlcPlayerView, autoplay: Bool?) in
                view.autoplay = autoplay ?? true
            }

            OnViewDidUpdateProps { (view: LibVlcPlayerView) in
                view.createPlayer()
            }

            AsyncFunction("play") { (view: LibVlcPlayerView) in
                view.play()
            }

            AsyncFunction("pause") { (view: LibVlcPlayerView) in
                view.pause()
            }

            AsyncFunction("stop") { (view: LibVlcPlayerView) in
                view.stop()
            }

            AsyncFunction("seek") { (view: LibVlcPlayerView, value: Double, type: String?) in
                view.seek(value, type ?? "position")
            }

            AsyncFunction("postAction") { (view: LibVlcPlayerView, action: Int) in
                view.postAction(action)
            }

            AsyncFunction("dismiss") { (view: LibVlcPlayerView) in
                view.dismiss()
            }
        }

        OnAppEntersBackground {
            MediaPlayerManager.shared.onAppBackground()
        }
    }
}

File: android/src/main/java/expo/modules/libvlcplayer/LibVlcPlayerModule.kt
package expo.modules.libvlcplayer

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.libvlcplayer.enums.AudioMixingMode
import expo.modules.libvlcplayer.enums.ContentFit
import expo.modules.libvlcplayer.records.Slave
import expo.modules.libvlcplayer.records.Tracks

private const val BUFFERING_EVENT = "onBuffering"
private const val PLAYING_EVENT = "onPlaying"
private const val PAUSED_EVENT = "onPaused"
private const val STOPPED_EVENT = "onStopped"
private const val END_REACHED_EVENT = "onEndReached"
private const val ENCOUNTERED_ERROR_EVENT = "onEncounteredError"
private const val TIME_CHANGED_EVENT = "onTimeChanged"
private const val POSITION_CHANGED_EVENT = "onPositionChanged"
private const val ES_ADDED_EVENT = "onESAdded"
private const val DIALOG_DISPLAY_EVENT = "onDialogDisplay"
private const val FIRST_PLAY_EVENT = "onFirstPlay"
private const val BACKGROUND_EVENT = "onBackground"

val playerEvents =
    arrayOf(
        BUFFERING_EVENT,
        PLAYING_EVENT,
        PAUSED_EVENT,
        STOPPED_EVENT,
        END_REACHED_EVENT,
        ENCOUNTERED_ERROR_EVENT,
        TIME_CHANGED_EVENT,
        POSITION_CHANGED_EVENT,
        ES_ADDED_EVENT,
        DIALOG_DISPLAY_EVENT,
        FIRST_PLAY_EVENT,
        BACKGROUND_EVENT,
    )

class LibVlcPlayerModule : Module() {
    override fun definition() =
        ModuleDefinition {
            Name("ExpoLibVlcPlayer")

            OnCreate {
                MediaPlayerManager.onModuleCreated(appContext)
            }

            OnDestroy {
                MediaPlayerManager.onModuleDestroyed()
            }

            View(LibVlcPlayerView::class) {
                Events(playerEvents)

                Prop("source") { view: LibVlcPlayerView, source: String? ->
                    view.source = source
                }

                Prop("options") { view: LibVlcPlayerView, options: ArrayList<String>? ->
                    view.options = options ?: ArrayList<String>()
                }

                Prop("tracks") { view: LibVlcPlayerView, tracks: Tracks? ->
                    view.tracks = tracks
                }

                Prop("slaves") { view: LibVlcPlayerView, slaves: ArrayList<Slave>? ->
                    view.slaves = slaves ?: ArrayList<Slave>()
                }

                Prop("scale") { view: LibVlcPlayerView, scale: Float? ->
                    view.scale = scale ?: DEFAULT_PLAYER_SCALE
                }

                Prop("aspectRatio") { view: LibVlcPlayerView, aspectRatio: String? ->
                    view.aspectRatio = aspectRatio
                }

                Prop("contentFit") { view: LibVlcPlayerView, contentFit: ContentFit? ->
                    view.contentFit = contentFit ?: ContentFit.CONTAIN
                }

                Prop("rate") { view: LibVlcPlayerView, rate: Float? ->
                    view.rate = rate ?: DEFAULT_PLAYER_RATE
                }

                Prop("time") { view: LibVlcPlayerView, time: Int? ->
                    view.time = time ?: DEFAULT_PLAYER_TIME
                }

                Prop("volume") { view: LibVlcPlayerView, volume: Int? ->
                    view.volume = volume ?: MAX_PLAYER_VOLUME
                }

                Prop("mute") { view: LibVlcPlayerView, mute: Boolean? ->
                    view.mute = mute ?: false
                }

                Prop("audioMixingMode") { view: LibVlcPlayerView, audioMixingMode: AudioMixingMode? ->
                    view.audioMixingMode = audioMixingMode ?: AudioMixingMode.AUTO
                }

                Prop("repeat") { view: LibVlcPlayerView, repeat: Boolean? ->
                    view.repeat = repeat ?: false
                }

                Prop("playInBackground") { view: LibVlcPlayerView, playInBackground: Boolean? ->
                    view.playInBackground = playInBackground ?: false
                }

                Prop("autoplay") { view: LibVlcPlayerView, autoplay: Boolean? ->
                    view.autoplay = autoplay ?: true
                }

                OnViewDestroys { view: LibVlcPlayerView ->
                    MediaPlayerManager.unregisterPlayerView(view)
                    view.destroyPlayer()
                }

                OnViewDidUpdateProps { view: LibVlcPlayerView ->
                    view.createPlayer()
                }

                AsyncFunction("play") { view: LibVlcPlayerView ->
                    view.play()
                }

                AsyncFunction("pause") { view: LibVlcPlayerView ->
                    view.pause()
                }

                AsyncFunction("stop") { view: LibVlcPlayerView ->
                    view.stop()
                }

                AsyncFunction("seek") { view: LibVlcPlayerView, value: Double, type: String? ->
                    view.seek(value, type ?: "position")
                }

                AsyncFunction("postAction") { view: LibVlcPlayerView, action: Int ->
                    view.postAction(action)
                }

                AsyncFunction("dismiss") { view: LibVlcPlayerView ->
                    view.dismiss()
                }
            }

            OnActivityEntersBackground {
                MediaPlayerManager.onAppBackground()
            }
        }
}
